docs/rfc/identity_resolution.md

```diff
@@ -0,0 +1,249 @@
+top-level entities (users), whose identity documents may themselves be signed by
+one or more top-level entities. Hence, in order to obtain the first repository
+and verifying its identity, we need to recurse indefinitely trying to resolve,
+clone, and verify the attesting entities' repositories.
+
+This is rather impractical from an implementation point of view: the initial
+clone has to block on a stack of asynchronous, high-latency network calls of
+(potentially) arbitrary depth, and since there is no availability guarantee nor
+-incentive, may still fail in the end, rendering the requested repository
+unusable.
+
+Hence, we are seeking to a. devise an incentive for the seeder to provide all
+data necessary to resolve and verify a repository, and b. to reduce latency by
+eliminating gossip queries and git fetches as much as possible.
+
+## Overview
+
+A lesser-known feature of the git suite are [namespaces], which are intended to
+provide network access (push/pull) to a _subset_ of a single repository on the
+server side. It turns out that this feature is not magical at all: a namespace
+is simply a `refs` hierarchy under the `namespaces` refs category of the
+"parent" repository. For example:
+
+    refs/namespaces/foobar/refs/heads/master
+
+The namespace is accessed by setting the `$GIT_NAMESPACE` environment variable,
+or passing the `--namespace` parameter to the `git` executable. However, the
+only commands which respect this are [git-upload-pack], [git-receive-pack], and
+[git-http-backend]. When invoked namespaced, they will simply only consider the
+`refs` hierarchy below the namespace (`foobar` in the example).
+
+The fact that no other git command seems to be aware of namespaces ([git-gc]
+would come to mind) is somewhat surprising, as it is otherwise not officially
+endorsed that one may create other categories apart from `heads`, `tags`, and
+`remotes` without any issues. Now that we know this, however, we can find
+interesting ways to exploit it in interesting and barely kosher ways.
+
+## Namespacing
+
+Consider a `radicle-link` peer would store **all** git repositories it is
+interested in in a single git repo, and made use of the namespaces feature to
+partition it into logical, smaller repos, which can be checked out individually.
+
+The namespacing scheme could look as follows:
+
+    $PEER_ID/refs/
+    `-- namespaces
+        `-- $IDENTITY
+            `-- refs
+                |-- heads # <-- code branches owned by $PEER_ID go here
+                |-- rad
+                |   |-- id # <-- points to the identity document history
+                |   |-- moi # <-- points to the identity of $PEER_ID
+                |   `-- ids
+                |       |-- $CERTIFIER_IDENTITY_1
+                |       `-- $CERTIFIER_IDENTITY_2
+                `-- remotes
+                    |-- $OTHER_PEER_ID
+                    |   `-- refs
+                    |       `-- heads
+                    `-- rad
```

Comment:
Expected this to be under `refs` like above. Is it not referring to the `$OTHER_PEER_ID`'s id and certifiers?

Reply:
Ya my ASCII art failed me